### PR TITLE
Add support for fetching rules from Rules Storage backend

### PR DESCRIPTION
**Pull request by onprem**
Following up on the discussion [here](https://github.com/observatorium/thanos-rule-syncer/pull/8#pullrequestreview-841868405), this PR adds support for fetching rules from Rules Storage backend directly. 

This is a non-breaking change and fetching rules from Observatorium API is still supported.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -103,14 +105,25 @@ func main() {
...
-
 	}
 
-	u, err := url.Parse(cfg.observatoriumURL)
-	if err != nil {
-		log.Fatalf("failed to parse Observatorium API URL: %v", err)
+	var u *url.URL
+
+	if cfg.rulesBackendURL != "" {
+		u, err := url.Parse(cfg.rulesBackendURL)
```
Comment (squat) [inline main.go]: when this is the case, we should really use the OpenAPI-generated client. There is no guarantee that the rules backend API and the Observatorium API will have the same interface, so we should adapt the `getRules` call appropriately

Comment (onprem) [reply main.go]: I have refactored the code to use a `fetcher` interface for all the rule fetching related logic.

File: fetcher.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	getRules(ctx context.Context) (rules io.ReadCloser, err error)
+}
+
+// observatoriumAPIFetcher fetches rules for a tenant from Observatorium API.
+type observatoriumAPIFetcher struct {
+	endpoint *url.URL
+	client   *http.Client
+}
+
+func newObservatoruimAPIFetcher(baseURL string, tenant string, client *http.Client) (*observatoriumAPIFetcher, error) {
```
Comment (matej-g) [inline fetcher.go]:
```suggestion
func newObservatoriumAPIFetcher(baseURL string, tenant string, client *http.Client) (*observatoriumAPIFetcher, error) {
```

File: main.go [hunk 3]
```diff
@@ -103,21 +103,30 @@ func main() {
...
+	var f fetcher
+
+	if cfg.rulesBackendURL != "" {
+		rulesFetcher, err := newRulesBackendFetcher(cfg.rulesBackendURL, client)
+		if err != nil {
+			log.Fatalf("failed to initialize Rules Backend fetcher: %v", err)
+		}
+		f = rulesFetcher
+	} else {
+		obsFetcher, err := newObservatoruimAPIFetcher(cfg.observatoriumURL, cfg.tenant, client)
```
Comment (matej-g) [inline main.go]:
```suggestion
		obsFetcher, err := newObservatoriumAPIFetcher(cfg.observatoriumURL, cfg.tenant, client)
```

**Review by matej-g — APPROVED** [review]
:medal_military:

**Review by squat — DISMISSED** [review]
LGTM % matej's comment
Good clean work @onprem

**Comment by squat** [comment]
Ah, hold up. Why isn't CI running? Let's hold off on merging until CI passes

**Review by squat — APPROVED** [review]
CI s running now
LGTM % green CI and Matej's comments

**Comment by squat** [comment]
@onprem this needs a regeneration of README.md and a rebase on main 🏄‍♀️

File: fetcher.go (hunk 2 above)
Comment (onprem) [reply fetcher.go]: Thanks for catching this!

**Comment by onprem** [comment]
> @onprem this needs a regeneration of README.md and a rebase on main surfing_woman

Done.

Thanks for the reviews, everyone :hugs:

**Pull request merged**